Libplanet/Net/Protocols/KBucket.cs

```diff
@@ -21,77 +26,110 @@ public KBucket(int size, Random random, ILogger logger)
             _size = size;
             _random = random;
             _logger = logger;
-            _peers = new List<(DateTimeOffset, BoundPeer)>();
+            _peers = new ConcurrentDictionary<BoundPeer, DateTimeOffset>();
             ReplacementCache = new List<BoundPeer>();
+            _bucketMutex = new AsyncLock();
 
             _lastUpdated = DateTimeOffset.UtcNow;
         }
 
         public int Count => _peers.Count;
 
-        // get most recently used peer.
-        public (DateTimeOffset, BoundPeer) Head =>
-            IsEmpty() ? (DateTimeOffset.MinValue, null) : _peers[_peers.Count - 1];
+        /// <summary>
+        /// Most recently used peer.
+        /// </summary>
+        public KeyValuePair<BoundPeer, DateTimeOffset> Head
+        {
+            get
+            {
+                var kvs = _peers.
+                    ToImmutableList()
+                    .Sort((kv1, kv2) => (int)(kv1.Value - kv2.Value).Ticks);
+                return kvs.LastOrDefault();
+            }
+        }
 
-        // get least recently used peer.
-        public (DateTimeOffset, BoundPeer) Tail =>
-            IsEmpty() ? (DateTimeOffset.MinValue, null) : _peers[0];
+        /// <summary>
+        /// Least recently used peer.
+        /// </summary>
+        public KeyValuePair<BoundPeer, DateTimeOffset> Tail
+        {
+            get
+            {
+                var kvs = _peers.
+                    ToImmutableList()
+                    .Sort((kv1, kv2) => (int)(kv1.Value - kv2.Value).Ticks);
```

Comment:
I guess we can use `_peers.Keys.Min()` instead of `.Sort()` and `FirstOrDefault()` combination.